python/lib/communication/dmod/communication/maas_request/ngen/ngen_exec_request_body.py

```diff
@@ -1,22 +1,45 @@
 from pydantic import Field, validator
 
+from .partial_realization_config import PartialRealizationConfig
+
 from dmod.core.meta_data import TimeRange
 from dmod.core.serializable import Serializable
 
 from typing import List, Optional
 
 
 class NGENRequestBody(Serializable):
+    """
+    Request body encapsulating data within outer request.
+
+    Encapsulated data to define a requested ngen job.  It includes details on the time range, hydrofabric, and
+    configurations need for executing ngen.  It may also include a reference to what forcing data to use.
+
+    An instance contains a reference to the ::class:`DataFormat.NGEN_JOB_COMPOSITE_CONFIG` dataset containing
+    configurations for the requested job.  In cases when this dataset doesn't yet exist, an instance also contains the
+    necessary details for generating such a dataset.  In particular, this includes:
+
+        - a realization config dataset id **OR** a ::class:`PartialRealizationConfig`
+        - (optionally) a BMI init config dataset id
+        - (optionally) a t-route configuration dataset id
+
+    When dataset ids are given, these are treated as sources for the new ::class:`DataFormat.NGEN_JOB_COMPOSITE_CONFIG`,
+    with the contents of the former copied into the latter as appropriate.
+    """
 
-    time_range: TimeRange
-    hydrofabric_uid: str
-    hydrofabric_data_id: str
-    realization_config_data_id: str = Field(description="Unique id of the realization config dataset for this request.")
-    forcings_data_id: Optional[str] = Field(None, description="Unique id of forcings dataset, if provided.")
-    bmi_config_data_id: str
+    time_range: TimeRange = Field(description="The time range over which to run ngen simulation(s).")
+    hydrofabric_uid: str = Field(description="The (DMOD-generated) unique id of the hydrofabric to use.")
+    hydrofabric_data_id: str = Field(description="The dataset id of the hydrofabric dataset to use.")
+    composite_config_data_id: str = Field(None, description="Id of required ngen composite config dataset.")
```

Comment:
The scenario I have in mind is when the composite config dataset does not already exist. If a request has non-None data_id values for the composite dataset _and_ one or more specialized source datasets, this implies the composite needs to be created from the data in the source(s) and giving the provided data_id. 

There could be a piece I am missing in the workflow for this all to come together, but that was what I was going for here. 

FWIW, I have a feeling we are going to need to overhaul the job and request types/workflow to build in more distinction. The combinatorics of how data versus data_ids are represented and impact execution paths has become too convoluted. But that can be it's own issue.

Reply:
> The scenario I have in mind is when the composite config dataset does not already exist. If a request has non-None data_id values for the composite dataset _and_ one or more specialized source datasets, this implies the composite needs to be created from the data in the source(s) and giving the provided data_id.

What are your thoughts on simplifying this and requiring that the dataset already exist? So, it would be prerequisite for some agent to create a composite dataset and send the id of that dataset along with their model exec request. It seems that we could the data service logic you've added in this PR to create / derive composite datasets, but this would just be a clear separation of concerns.

Reply:
FWIW, this might be better suited to a larger refactoring effort like you alluded to in your comment.

Reply:
> What are your thoughts on simplifying this and requiring that the dataset already exist?

Yeah, the short version is that it may not actually simplifying things.

The request types are designed to not assume the dataset exists.  They can potentially include explicit, user-supplied configuration data in places like the _PartialRealizationConfig_ and/or several of the attributes in _NgenCalRequestEstimationConfig_.  I suppose the same idea applies with implicit config data and source dataset ids.  

The job workflow is also built to not assume datasets exist.  It includes a step expressly for generating any new, required datasets (for which such generation is supported).  And, at the earlier step of determining whether it has all the data needed for the job, it will (when needed) test if sufficient data is available for generating a dataset for an otherwise unfulfilled requirement.

We could separate the behavior for generating datasets from the job workflow and the job request types.  Perhaps we _should_ do this, especially if we eventually want robust CLI and API capabilities.  But that would touch a lot of places.  And further, if the expectation is to provide a GUI form where the user seamlessly both explicitly configures and starts a job, that would shift responsibility to the GUI for handling a good bit of behind the scenes work to create and coordinate dataset creation requests with job requests.

Reply:
Thanks for providing context and background, @robertbartel! I think I went a little too far in my last comments. We have fairly good separation of concerns behind this layer that I was neglecting to consider. I _dont_ think that we want to separate the behavior like I mentioned before. I do however think we should consider how to reduce the combinatorics of potential field combinations in this request model. So for example, splitting `NGENRequestBody` into two models one that supports only a single required `composite` dataset id and another that supports providing optional routing, realization, and bmi config dataset ids. I think this will make it more clear from the caller side what combinations of fields are allowed. Thoughts?

Reply:
> I think this will make it more clear from the caller side what combinations of fields are allowed.

Does it?  The spinoff type for an existing composite config would slightly simplify the easiest scenario - leave out everything that is optional - by instead excluding things by design.  The spinoff type that handled creating a new composite dataset would basically need to be the same as the current class, unless we further separated that spinoff into different types for providing the actual config data versus providing source data_id references.

And, we'd have to do that all this again for the calibration request body type.

If the problem is ambiguity on the calling side, we could more easily add factory class methods to make the initialization paths more clear.  We could also add some additional properties that indicated in a more summary fashion the different types of state represented.

Reply:
> Does it? The spinoff type for an existing composite config would slightly simplify the easiest scenario - leave out everything that is optional - by instead excluding things by design. The spinoff type that handled creating a new composite dataset would basically need to be the same as the current class, unless we further separated that spinoff into different types for providing the actual config data versus providing source data_id references.

Maybe it does not _simplify_ things, but it does make the possible request variants more explicit through using types instead of logic. I think this adds a level of clarity and maintainability that is desirable. However, as you mentioned previously, we probably need to rethink the model exec request structures in the future anyways, so it might not be worth the effort.

My main concern is that we end up with a catchall ngen request type that requires a lot of logic spread across the codebase to enforce what combination of fields are valid and we can handle. This same logic would bleed into client code and need to be updated accordingly. At face value, it seems that it would easier, although more verbose, to treat each combination of fields as its own type. This should reduce the amount of imperative code required to enforce invariants, simplify introduction of new request type, and make it more explicit _what_ combination of fields are accepted.

Reply:
> My main concern is that we end up with a catchall ngen request type that requires a lot of logic spread across the codebase to enforce what combination of fields are valid and we can handle. 

Yeah, I agree.  My main concern is that separating the types would only shuffle this the logic to different places, and not reduce it into the fundamentals of the design.  I could be wrong it this, but regardless, I think that ends up being part of an otherwise-alluded-to larger overhaul.